src/Workspaces/Remote/ServiceHub/Host/RemoteWorkspace.cs

```diff
@@ -114,19 +142,21 @@ public AssetProvider CreateAssetProvider(Checksum solutionChecksum, SolutionAsse
             // increment the in-flight of that solution until we decrement it at the end of our try/finally block.
             var (inFlightSolution, solutionTask) = await AcquireSolutionAndIncrementInFlightCountAsync().ConfigureAwait(false);
 
+            Exception? exception = null;
             try
             {
-                return await ProcessSolutionAsync(inFlightSolution, solutionTask).ConfigureAwait(false);
+                var solution = await GetSolutionAsync(inFlightSolution, solutionTask).ConfigureAwait(false);
+                return new PinnedSolution(this, inFlightSolution, solution);
             }
-            catch (Exception ex) when (FatalError.ReportAndPropagateUnlessCanceled(ex, cancellationToken, ErrorSeverity.Critical))
+            catch (Exception ex) when ((exception = ex) == null)
```

Comment:
What is the == null here? Or intentionally a way to force the block to always be false? This is missing an explanatory comment.

Reply:
(see earlier comment being unsure why we have a catch/finally at all)

Reply:
yes.  this is a pattern we have in OOP that tomas likes for ensuring callstacks remain good.